File: apps/rails/spec/factories/company_roles.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :company_role do
    company
    name { Faker::Job.title }
    job_description { Faker::Quote.famous_last_words }
    capitalized_expense { Faker::Number.between(from: 0, to: 80) }

    transient do
      pay_rate_in_subunits { Faker::Number.between(from: 100_00, to: 200_00) }
      pay_rate_type { CompanyRoleRate.pay_rate_types[:hourly] }
    end

    factory :project_based_company_role do
      name { "Project-based Engineer" }
      pay_rate_type { CompanyRoleRate.pay_rate_types[:project_based] }
    end

    factory :salary_company_role do
      name { "Salaried Engineer" }
      pay_rate_type { CompanyRoleRate.pay_rate_types[:salary] }
    end

    after :build do |company_role, evaluator|
      company_role.rate = build(:company_role_rate, pay_rate_in_subunits: evaluator.pay_rate_in_subunits,
                                                    pay_rate_type: evaluator.pay_rate_type,
                                                    company_role: nil)
    end
  end
end

File: e2e/factories/companyRoles.ts
import { faker } from "@faker-js/faker";
import { db } from "@test/db";
import { companiesFactory } from "@test/factories/companies";
import { PayRateType } from "@/db/enums";
import { companyRoleRates, companyRoles } from "@/db/schema";
import { assert } from "@/utils/assert";

export const companyRolesFactory = {
  create: async (
    overrides: Partial<typeof companyRoles.$inferInsert> = {},
    rateOverrides: Partial<typeof companyRoleRates.$inferInsert> = {},
  ) => {
    const [role] = await db
      .insert(companyRoles)
      .values({
        companyId: overrides.companyId || (await companiesFactory.create()).company.id,
        name: faker.person.jobTitle(),
        jobDescription: "", // Empty string but still required by schema
        capitalizedExpense: faker.number.int({ min: 0, max: 80 }),
        ...overrides,
      })
      .returning();
    assert(role != null);

    const [rate] = await db
      .insert(companyRoleRates)
      .values({
        companyRoleId: role.id,
        payRateInSubunits: faker.number.int({ min: 10000, max: 20000 }),
        payRateType: PayRateType.Hourly,
        ...rateOverrides,
      })
      .returning();

    return { role, rate };
  },

  createProjectBased: async (
    overrides: Partial<typeof companyRoles.$inferInsert> = {},
    rateOverrides: Partial<typeof companyRoleRates.$inferInsert> = {},
  ) =>
    companyRolesFactory.create(
      { name: "Project-based Engineer", ...overrides },
      { payRateType: PayRateType.ProjectBased, ...rateOverrides },
    ),

  createSalaried: async (
    overrides: Partial<typeof companyRoles.$inferInsert> = {},
    rateOverrides: Partial<typeof companyRoleRates.$inferInsert> = {},
  ) =>
    companyRolesFactory.create(
      { name: "Salaried Engineer", ...overrides },
      { payRateType: PayRateType.Salary, ...rateOverrides },
    ),
};
